### PR TITLE
runtime heatmap generation

### PR DESCRIPTION
**Pull request by carns**
This PR introduces an auto-scaling heatmap histogram at runtime; see darshan-heatmap.c.

Multiple APIs can register with it and keep independent histograms (each histogram is a separate Darshan record).  All of the histogram records are stored in a single dedicated histogram module from a log perspective.

PR currently just has the runtime library portion implemented, and it is only hooked up to the POSIX module.  Need to:

- [x] implement logutils portion
- [x] add support for variable number of bins in log format
- [x] test performance (probably repeat tests from recent ECP experiments)
- [x] implement (or at least stub in) the agg_records() logutils function for future use
- [x] documentation
- [x] squash commits
- [x] hook into mpi-io and stdio modules

### CODE REVIEW

**Comment by carns** [comment]
implemented enough of logutils for darshan-parser to work; going to skip ahead now to runtime overhead measurements.  Current prototype is connected to posix, mpiio, and stdio modules.

**Comment by carns** [comment]
This PR is meant to address #468

**Comment by carns** [comment]
Complete now.

Fixes #468

**Comment by carns** [comment]
This PR does not provide Python bindings for the heatmap data.  That will need to be added as a separate PR after this one lands.

**Review by shanedsnyder — CHANGES_REQUESTED** [review]

**Comment by carns** [comment]
Thanks @shanedsnyder , all good feedback.  It looks like I botched a late commit that got squashed into this that contributed to the problems (the broken lock macro at least), but I'll get it untangled shortly.

File: darshan-runtime/lib/darshan-heatmap.c
Comment (carns) [reply darshan-runtime/lib/darshan-heatmap.c]: Doh!  pushed fix just now.  The underlying root cause was that the stdatomic path was accidentally disabled in Darshan, fixed in #605 , which made this obvious cut and paste error show up here after rebasing.

File: darshan-runtime/lib/darshan-heatmap.c
Comment (carns) [reply darshan-runtime/lib/darshan-heatmap.c:210]: Yep, good catch.  Fixed in c50c08cb34b25148f736cabef086c0f28b12f6ab.

File: darshan-util/darshan-heatmap-logutils.c
Comment (carns) [reply darshan-util/darshan-heatmap-logutils.c:149]: yep, confirmed; this bug makes darshan-convert fail to generate a valid darshan file if it has heatmap data in it.  Will work on a fix shortly.

**Comment by jakobluettgau** [comment]
Just to confirm: The last few commits let me successfully create a darshan log with a heatmap in it ;)

File: darshan-util/darshan-heatmap-logutils.c
Comment (carns) [reply darshan-util/darshan-heatmap-logutils.c]: fixed in 4656ad66b198a0197cf9b3eb555d53572fc17aea.  I'm making a note to think about how to unit test for this sort of thing so it's harder to miss...

File: darshan-runtime/lib/darshan-heatmap.c
Comment (carns) [reply darshan-runtime/lib/darshan-heatmap.c]: Good point.  There isn't a great way to automatically track how many modules might use heatmaps (I'm worried we would forget to update the value) so I just picked a separate #define constant for now and flagged it to make configurable in the future.  It is set to 8 for now, and will not inadvertently escalate if MAX_MODS is increased.  Fixed in  684fc14e60361874b6b31c5ac7a21f488e78973c.

File: darshan-runtime/lib/darshan-heatmap.c
Comment (carns) [reply darshan-runtime/lib/darshan-heatmap.c:169]: That's interesting.  I got tripped up here by being over-reliant on address sanitizer, which apparently doesn't complain because the pointers are all technically still reachable at exit.  I switched to valgrind `valgrind --leak-check=full --show-leak-kinds=all` and now I see the leaks there.  Will fix.

Comment (carns) [reply darshan-runtime/lib/darshan-heatmap.c:169]: Fixed in ccefac9d67703efe29da8e6fd05e93482dbd1e0e

**Comment by carns** [comment]
> Just to confirm: The last few commits let me successfully create a darshan log with a heatmap in it ;)

Excellent, thanks for the confirmation :) I'm not sure how I ended up with such a broken PR but I think it's in better shape now.

**Comment by nawtrey** [comment]
@carns confirmed over PM that this will generate data with shape `xbins x ybins`, where `ybins <= nprocs` (ranks without activity are not listed). This is consistent with the methods in `heatmap_handling.py`.  

I looked over a couple logs provided to me by @jakobluettgau using `darshan-parser` built from this branch. It appears all the data is there, and it should be possible to manipulate it into a form consistent with the current heatmap dataframe. So as far as the data is concerned, LGTM.

**Comment by carns** [comment]
> So as far as the data is concerned, LGTM.

Thanks @nawtrey !

**Review by shanedsnyder — APPROVED** [review]

**Comment by shanedsnyder** [comment]
All looks good to me, thanks for cleaning that stuff up @carns! Feel free to merge this in if you guys aren't tying up any more loose ends.

**Comment by tylerjereddy** [comment]
I think Phil may add an option to disable as safety precaution, but we were hoping to merge soon-ish then backport it yeah.

**Review by roblatham00 — APPROVED** [review]
Took a quick scan through this and it looks right.

**Pull request merged**
